pydantic/_internal/_typing_extra.py

```diff
@@ -85,11 +85,27 @@ def is_literal(tp: Any, /) -> bool:
     return _is_typing_name(get_origin(tp), name='Literal')
 
 
-# TODO remove and replace with `get_args` when we drop support for Python 3.8
-# (see https://docs.python.org/3/whatsnew/3.9.html#id4).
 def literal_values(tp: Any, /) -> list[Any]:
-    """Return the values contained in the provided `Literal` special form."""
+    """Return the values contained in the provided `Literal` special form.
+
+    If one of the literal values is a PEP 695 type alias, recursively parse
+    the type alias' `__value__` to unpack literal values as well. This function
+    *doesn't* check that the type alias is referencing a `Literal` special form,
+    so unexpected values could be unpacked.
+    """
+    # TODO When we drop support for Python 3.8, there's no need to check of `is_literal`
+    # here, as Python unpacks nested `Literal` forms in 3.9+.
+    # (see https://docs.python.org/3/whatsnew/3.9.html#id4).
     if not is_literal(tp):
+        # Note: we could also check for generic aliases with a type alias as an origin.
+        # However, it is very unlikely that this happens as type variables can't appear in
+        # `Literal` forms, so the only valid (but unnecessary) use case would be something like:
+        # `type Test[T] = Literal['a']` (and then use `Test[int]`).
```

Comment:
Did you mean to use a str literal here?

Reply:
Nope, if you do:

```python
type Test[T] = Literal['whatever']
MyLit = Literal[Test[int]]
```

`MyLit` is a `GenericAlias` instance, and here we only check for `is_type_alias_type`. But as I mentioned in the comment it is very unlikely to encounter such use cases, as it is not necessary (the type variable is useless here as you can't do something like `type Test[T] = Literal[T]`).